Dynamo: Fix graph break when iterating over tensor

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94350
* __->__ #94326

Supports the following with dynamic shapes:
```python
for element in tensor:
    # do stuff with element
```

Approach follows what's done when `call_range()` is invoked with dynamic shape inputs: guard on tensor size and continue tracing with a real size value from `dyn_dim0_size.evaluate_expr()`.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

jbschlosser committed:
Dynamo: Fix graph break when iterating over tensor

[ghstack-poisoned]

voznesenskym commented on test/dynamo/test_misc.py:
```diff
@@ -311,6 +311,17 @@ def fn(x):
         r2 = opt_fn(i)
         self.assertTrue(same(r1, r2))
 
+    def test_tensor_iter(self):
+        def fn(x):
+            for y in x:
+                y.add_(1.0)
+            return y
+
+        # expect extra size node for dynamic
+        torch._dynamo.testing.standard_test(
+            self, fn, 1, expected_ops=20, expected_ops_dynamic=21
+        )
```
can you add a guard failure test? `evaluate_expr` below incurs a guard, which is what we want, but I would like to see something like:

1) trace fn with X
2) run fn with Y where Y.size != X.size
3) run fn with Z where Z.size != Y.size != X.size

A) the guard failure fn should report that a guard failed because the value from Z is not the value from Y
B) we should also add comparisons to real values between 2 and 3 above to ensure we are not baking in /compiling in an unwound loop of a fixed size

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -217,15 +217,25 @@ def try_generic_attr_handling():
-                return super(TensorVariable, self).unpack_var_sequence(tx)
-        options = VariableTracker.propagate(self)
+                dyn_length = self.call_method(
+                    tx, "size", [ConstantVariable(0, **options)], {}
```
There isn't really point in propagating the self options into the fresh ConstantVariable. ConstantVariable(0) is fine.

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -217,15 +217,25 @@ def try_generic_attr_handling():
         return result
 
+    def has_unpack_var_sequence(self, tx):
+        return (self.size is not None and len(self.size) > 0) or (
```
Can't you technically unpack the sequence even if it has no elements?

ezyang replied:
Yeah, scalar is a different thing; your code will error here too because size(0) will fail

jbschlosser replied:
I think I'm doing it right by checking `len(size)`, no?
* scalar case: `len(size) == 0`; don't try to unpack it
* empty sequence: `len(size) == 1` with `size(0) == 0`; okay to unpack

ezyang replied:
Oh oopsies, you are right!

ezyang commented:
See comments but o/w looks reasonable

jbschlosser committed:
Update on "Dynamo: Fix graph break when iterating over tensor"


Supports the following with dynamic shapes:
```python
for element in tensor:
    # do stuff with element
```

Approach follows what's done when `call_range()` is invoked with dynamic shape inputs: guard on tensor size and continue tracing with a real size value from `dyn_dim0_size.evaluate_expr()`.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser committed:
Update on "Dynamo: Fix graph break when iterating over tensor"


Supports the following with dynamic shapes:
```python
for element in tensor:
    # do stuff with element
```

Approach follows what's done when `call_range()` is invoked with dynamic shape inputs: guard on tensor size and continue tracing with a real size value from `dyn_dim0_size.evaluate_expr()`.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
